### PR TITLE
WIP Fix issues with assign/link widget in ZUIPersonEditGridCell

### PR DESCRIPTION
**Pull request by rebecarubio**
## Description
This PR fixes different issues related to the performance of the ZUIPersonEditCell component.

## Screenshots
1- Widget in Submission list 

https://user-images.githubusercontent.com/36491300/228483686-7d6880ec-4391-4858-97ed-2d6fc223571d.mp4

2- Widget in Views

https://user-images.githubusercontent.com/36491300/228485177-f6b963c0-31f4-461c-83ba-91c1da7544e4.mp4

## Changes
* Adds the `first_name `and `last_name ` of the respondent as a new query parameter to search for `suggestedPeople` to be able to display the PersonHoverCard in those persons that doesn´t have an email settled.
* Adds a `CircleProgress `component to show to the user that the searching action is not completed.
* Adds padding to the edit cell to make obvious that the user can type and search
* Removes extra paddingTop in the `Popper `element
* Fixes the sticky position of the `subheader` when scrolling.
* Adds filtering in the `suggested people` list to check against the input values
* Adds a "keep typing.." state 
* Adds a "no matching results" message if the results of the search are 0
* In Submission List, if the respondent is anonymous, we don´t show suggested people.

## Notes to reviewer
None

## Related issues
Relates to #1067

### CODE REVIEW

File: src/features/surveys/components/SurveySubmissionsList.tsx [hunk 1]
```diff
@@ -157,20 +161,28 @@ const SurveySubmissionsList = ({
...
+      row.respondent?.email ||
+      row.respondent?.first_name ||
+      row.respondent?.last_name ||
+      '';
     let { data: suggestedPeople } = useQuery(
-      ['peopleSearchResults', email],
-      getPeopleSearchResults(email, orgId as string),
+      ['peopleSearchResults', emailOrName],
+      getPeopleSearchResults(emailOrName, orgId as string),
       {
```
Comment (ziggabyte) [inline src/features/surveys/components/SurveySubmissionsList.tsx:179]: I assum what's happening here is that I'm getting three "Casey" because they get included one time for each search criteria (first name, last name and email)? So iI think this needs to be filtered down to avoid duplicates
bild

File: src/zui/ZUIPersonGridEditCell.tsx [hunk 2]
```diff
@@ -96,8 +97,10 @@ const ZUIPersonGridEditCell: FC<{
           inputProps={autoComplete.getInputProps()}
           onChange={() => setSearching(true)}
           placeholder={messages.personSelect.search()}
+          sx={{ paddingLeft: '10px' }}
```
Comment (ziggabyte) [inline src/zui/ZUIPersonGridEditCell.tsx:100]: This made me so calm haha i got stressed out by the lack of padding before <3

File: src/zui/ZUIPersonGridEditCell.tsx [hunk 3]
```diff
@@ -195,12 +198,20 @@ const ZUIPersonGridEditCell: FC<{
...
+                        {...autoComplete.getListboxProps()}
+                        subheader={
+                          <ListSubheader sx={{ position: 'relative' }}>
+                            {suggestedPeople.length >= 0 &&
+                              messages.personGridEditCell.searchResults()}
+                          </ListSubheader>
+                        }
+                      >
+                        {searchResults.length == 0 && (
+                          <CircularProgress
```
Comment (ziggabyte) [inline src/zui/ZUIPersonGridEditCell.tsx:211]: I think the criteria for when it shows the CircularProgress can be improved. With this criteria, it does not differentiate between "no search results" and "loading/searching" and i felt that it was a bit misleading, since the spinner usually means "wait a second, things will appear here". But if there are no search results it will just spin forever. 
Could there be a way to differentiate between "i have stuff that's gonna appear here" and "no search results"? 
bild

File: src/features/surveys/components/SurveySubmissionsList.tsx [hunk 4]
```diff
@@ -191,7 +203,7 @@ const SurveySubmissionsList = ({
         cell={row.respondent}
         onUpdate={updateCellValue}
         removePersonLabel={messages.submissions.unlink()}
-        suggestedPeople={suggestedPeople}
+        suggestedPeople={respondents}
         suggestedPeopleLabel={messages.submissions.suggestedPeople()}
```
Comment (ziggabyte) [inline src/features/surveys/components/SurveySubmissionsList.tsx:207]: The list of suggested people (respondents) is unaffected by the search string, is there a specific reason for this? I found it a bit clunky to search and not have the suggested people be filtered by my search string.

File: src/features/surveys/components/SurveySubmissionsList.tsx [hunk 5]
```diff
@@ -157,20 +161,28 @@ const SurveySubmissionsList = ({
...
       {
-        enabled: email.length >= 2,
+        enabled: emailOrName.length >= 2,
         retry: true,
       }
     );
 
-    if (!suggestedPeople) {
+    if (!suggestedPeople || suggestedPeople === undefined) {
       suggestedPeople = [];
```
Comment (ziggabyte) [inline src/features/surveys/components/SurveySubmissionsList.tsx:186]: isn't `undefined` falsy? I assume there is a reason for this addition but it made me curious.

File: src/zui/ZUIPersonGridEditCell.tsx [hunk 6]
```diff
@@ -201,7 +201,7 @@ const ZUIPersonGridEditCell: FC<{
                       <List
                         {...autoComplete.getListboxProps()}
                         subheader={
-                          <ListSubheader>
+                          <ListSubheader sx={{ position: 'relative' }}>
                             {suggestedPeople.length >= 0 &&
```
Comment (ziggabyte) [inline src/zui/ZUIPersonGridEditCell.tsx:205]: The ListSubheader on line 185 also needs the same positioning! 
Also, I think there's a bit too much space above this header? I can't find exactly where this padding/margin is set, but it feels slightly off compared to other components

**Review by ziggabyte — CHANGES_REQUESTED** [review]
Nice! I commented on all the things I noticed when playing with it! 

Also, regarding "Jane Respondent", it seems lik this user has some property that makes it impossible to link her. Is this a case of strange dummy data, or does she have some set of properties that a real user could realistically have?

**Comment by niklasva82** [comment]
I'm actually not so sure it's a good idea to search for first and last name, there can be so many people with the same name in a large organization, I think that would encourage linking the wrong person. If the email doesn't match any person object, the organizer can still manually search for the person with the understanding that the results they get may not necessarily be that person.

File: src/features/surveys/l10n/messageIds.ts [hunk 7]
```diff
@@ -152,7 +152,7 @@ export default makeMessages('feat.surveys', {
     lastNameColumn: m('Last name'),
     link: m('Link'),
     personRecordColumn: m('Respondent'),
-    suggestedPeople: m('Suggested people'),
+    suggestedPeople: m('Other linked respondents'),
```
Comment (niklasva82) [inline src/features/surveys/l10n/messageIds.ts:155]: Why this change? We're looking for people who match the specific person here, not other respondents as in the case of views?

**Comment by ziggabyte** [comment]
> I'm actually not so sure it's a good idea to search for first and last name, there can be so many people with the same name in a large organization, I think that would encourage linking the wrong person. If the email doesn't match any person object, the organizer can still manually search for the person with the understanding that the results they get may not necessarily be that person. This is also how it works in gen 2.

My interpretation of the problem this is trying to solve is that if you wanted to link a user that doesn't have an email to their profile, they were impossible to search for here since it based the search only searched email adresses...

**Comment by niklasva82** [comment]
> > I'm actually not so sure it's a good idea to search for first and last name, there can be so many people with the same name in a large organization, I think that would encourage linking the wrong person. If the email doesn't match any person object, the organizer can still manually search for the person with the understanding that the results they get may not necessarily be that person. This is also how it works in gen 2.
> 
> My interpretation of the problem this is trying to solve is that if you wanted to link a user that doesn't have an email to their profile, they were impossible to search for here since it based the search only on email adresses...

I understand what problem it's trying to solve, but I think if the user is provided with someone with simply the same name, when they are used to suggestions based on email, they just "by automation" pick the suggested person. If they are forced to search for the name, they will be more aware of the fact that this could be a duplicated name. This is how I work when I link many submissions in gen 2, suggested people are almost autopilot for me.

**Comment by niklasva82** [comment]
> > I'm actually not so sure it's a good idea to search for first and last name, there can be so many people with the same name in a large organization, I think that would encourage linking the wrong person. If the email doesn't match any person object, the organizer can still manually search for the person with the understanding that the results they get may not necessarily be that person. This is also how it works in gen 2.
> 
> My interpretation of the problem this is trying to solve is that if you wanted to link a user that doesn't have an email to their profile, they were impossible to search for here since it based the search only on email adresses...

Ah wait, I may have written a response before understanding what you were saying :D As far as I know it was still possible to search for names, just that the suggestion did not search for names.

File: src/features/surveys/components/SurveySubmissionsList.tsx (hunk 1 above)
Comment (rebecarubio) [reply src/features/surveys/components/SurveySubmissionsList.tsx:183]: Good catch! I´m filtering now to avoid duplicates :D

File: src/features/surveys/components/SurveySubmissionsList.tsx (hunk 4 above)
Comment (rebecarubio) [reply src/features/surveys/components/SurveySubmissionsList.tsx:211]: The filtering of the suggested people is happening in the `ZUIPersonEditGridCell `and I fix it to make it work :)

File: src/zui/ZUIPersonGridEditCell.tsx (hunk 3 above)
Comment (rebecarubio) [reply src/zui/ZUIPersonGridEditCell.tsx:243]: You are right! I´ve added a message to make clear when there are no results!

File: src/features/surveys/l10n/messageIds.ts (hunk 7 above)
Comment (rebecarubio) [reply src/features/surveys/l10n/messageIds.ts]: You are right, I changed it again! I misunderstood what the suggested people was! thanks for pointing out!

**Comment by rebecarubio** [comment]
Thanks for the feedback @ziggabyte and @niklasva82! I´ve made changes trying to include your ideas so I would like if you can take a look at it now :)

**Review by ziggabyte — COMMENTED** [review]
My comments to this version: 

If I type something that seems to have no search matches, the popup does not open. My expectation was that it would open and say "No search results" (or similar message).
bild

I don't get "Suggested people" when I want to link an anonymous submission. My expectation was that I would get suggested people, based on email adresses. If I type into the text field of a linked submission, I get "Suggested people" based on email adresses, as expected. I thought the two should behave the same?
bild

**Comment by rebecarubio** [comment]
> My comments to this version:
> 
> If I type something that seems to have no search matches, the popup does not open. My expectation was that it would open and say "No search results" (or similar message). bild
> 
> I don't get "Suggested people" when I want to link an anonymous submission. My expectation was that I would get suggested people, based on email adresses. If I type into the text field of a linked submission, I get "Suggested people" based on email adresses, as expected. I thought the two should behave the same? bild

it´s true that if the respondent it´s anonymous the popover doesn´t open, thanks! I´ll fix it!
And if I remembered correctly we said that we will not suggest people if the respondent was anonymous (in this case we don´t know their email also)

**Comment by ziggabyte** [comment]
> it´s true that if the respondent it´s anonymous the popover doesn´t open, thanks! I´ll fix it!
> And if I remembered correctly we said that we will not suggest people if the respondent was anonymous (in this case we don´t > know their email also)

Aaah ofcourse it's not possible to suggest people based on email if they are anonymous! Forget my comment about that :)

**Review by richardolsson — APPROVED** [review]
This is working ok now, so I think it's fine to merge it.

I did react to the loading indicator not showing up right away, but after a short delay. I'm pretty sure that is caused by the debouncing, that will need to be modified anyway once we rebuild this with a model/repo/store integration, so it's not worth spending time on now.
